### DISCUSSION TITLE
Theia difference between frontend and backend plugin

### DISCUSSION THREAD

**Discussion by Rmttf** [eclipse-theia/theia#11001] [Q&A] [answered]
I didn't find more descriptions of thea's front-end plugins. I compared the front-end plug-ins generated by the code tool *(@theia/generator-plugin)* with the back-end plug-ins, and it doesn't look like much difference.

1. Can you introduce the difference between front-end and back-end plug-ins? 
2. Why distinguish between front-end plug-ins and back-end plug-ins?

**Comment by vince-fugnitto** [comment, marked as answer]
@Rmttf thank you for the discussion, hopefully the following helps https://theia-ide.org/docs/authoring_plugins/#design:

> There are two natures of plug-ins:
> 
> - **Backend plug-in**: If you're familiar with VS Code extensions, it's very close. The plug-in's code is running in its own process on the  server side. API is called and it's the API that will send some actions on user's browser/UI to register new commands, etc. All the callbacks are executed on the server side on a dedicated process.
> - **Frontend plug-in**: In that case, callbacks are executed in a worker thread on the UI/browser. These plug-ins are only authorized to   use "browser compliant" modules. For example opening or writing to a file is impossible as all the code of the plug-in is running on the browser side. But this approach is helpful if you really want to have some stuff on the client side to avoid some network operations.

**Reply by Rmttf** [reply]
Thanks for the reply. I don't know if I understand it correctly: the difference between the two types of plugins is only the environment in which the plugin code runs. The frontend plug-in runs in the browser's worker, and the backend plug-in runs on the thread of the remote service.

**Reply by vince-fugnitto** [reply]
Right, so based on the documentation, the simplest definition would be:

- **backend plugin**: 
  - runs in a separate process on the **server** (backend) side
  - possibility to interact with the browser side through APIs


- **frontend plugin**:
  - runs in the **browser** (frontend) side
  - can only interact with the frontend/browser side, it is not possible to perform or execute backend functionality
  - limited API (since its frontend specific)

Please note that in most cases it is probably simpler to create a **vscode extension** and consume it in a theia application than write a **theia plugin**. This also gives you the possibility to target a wider range of audiences for different products that support the VS Code API.

More information about the differences between **theia extensions**, **vscode extensions** and **theia plugins** can be viewed here:
- https://theia-ide.org/docs/extensions/#extensions-and-plugins
